### ISSUE TITLE
why using fixed value transition matrix can help keeping the scores for tag sequences including outer entities higher than those of tag sequences including inner entities?

### ISSUE THREAD

**Issue by wangwang110** [yahshibu/nested-ner-2019-bert#4]
2.1 Usage of CRF 

 "that each element of the transition matrix of each CRF has a fixed value according to whether it corresponds to a legal transition (e.g., B-X to I-X in IOBES tagging scheme, where X is the name of entity type) or an illegal one (e.g., O to I-X). This
is helpful for keeping the scores for tag sequences including outer entities higher than those of tag sequences including inner entities"

using the fingure 2 as example， according to transition matrix ，I  think the following two tag sequences has the same score.
[s] o o o o o B-P I-P I-P E-P o o ... o o [E]
[s] o o o o o o o B-P E-P o o ... o o [E]
I wonder how the  fixed value transition matrix help keeping the scores for tag sequences including outer entities higher

**Comment by yahshibu** [comment]
Thank you for having an interest in our work!

Our explanation is not complete in Section 2.1. We couldn't put a full explanation there due to the limitation of space. The mechanism is described across Sections 2.1, 2.2, and 2.3.

Fixing values doesn't **directly** help to keep the scores for tag sequences including outer entities higher than those of tag sequences including inner entities. Instead, fixing values guarantees that the best path at the 2nd level is the same as the best path at the 1st level, that the best path at the 3rd level is the same as the second-best path at the 2nd level, that the best path at the 4th level is the same as the second-best path at the 3rd level, and so on. Fixing values ensures these consistent relationships.

It is our objective function for training that makes the scores for tag sequences including outer entities higher than those of tag sequences including inner entities. But, the objective function itself doesn't help to keep consistent relationships across different levels during inference.

In summary, fixing values assists the objective function by maintaining consistent relationships during inference.

**Comment by wangwang110** [comment]
> Thank you for having an interest in our work!
> 
> Our explanation is not complete in Section 2.1. We couldn't put a full explanation there due to the limitation of space. The mechanism is described across Sections 2.1, 2.2, and 2.3.
> 
> Fixing values doesn't **directly** help to keep the scores for tag sequences including outer entities higher than those of tag sequences including inner entities. Instead, fixing values guarantees that the best path at the 2nd level is the same as the best path at the 1st level, that the best path at the 3rd level is the same as the second-best path at the 2nd level, that the best path at the 4th level is the same as the second-best path at the 3rd level, and so on. Fixing values ensures these consistent relationships.
> 
> It is our objective function for training that makes the scores for tag sequences including outer entities higher than those of tag sequences including inner entities. But, the objective function itself doesn't help to keep consistent relationships across different levels during inference.
> 
> In summary, fixing values assists the objective function by maintaining consistent relationships during inference.

thank you for your reply，I understand.  
another question, did you do experiment to compare the performance of using multi-crfs and using one crf

**Comment by yahshibu** [comment]
No, we didn't.

We have three reasons why we didn't. The first reason is that, if a single CRF was used for all entity types, the decoder could not always narrow down spans with the 2nd best path. The 2nd best path in a single CRF could result in the same span tagged a different entity type such as

[s] o o o o o B-P I-P I-P E-P o o ... o o [E]
~[s] o o o o o B-C I-P I-P E-P o o ... o o [E]~
[s] o o o o o B-C I-C I-C E-C o o ... o o [E]

. We would have to watch lower-ranked paths. The second reason is that there are several cases where the same mention span is assigned multiple entity types ([link](https://github.com/yahshibu/nested-ner-tacl2020-transformers/issues/3)). In such cases, we cannot decide which tag sequence should be ranked higher. By using separate CRFs for different entity types, we don't need to consider which should be ranked higher. The third reason is that [Muis and Lu (2017)](https://www.aclweb.org/anthology/D17-1276/) demonstrated that this approach of multiple CRFs would perform better on nested NER datasets and even a flat NER dataset than the standard approach of a single CRF for all entity types.

**Comment by wangwang110** [comment]
> No, we didn't.
> 
> We have three reasons why we didn't. The first reason is that, if a single CRF was used for all entity types, the decoder could not always narrow down spans with the 2nd best path. The 2nd best path in a single CRF could result in the same span tagged a different entity type such as
> 
> ```
> [s] o o o o o B-P I-P I-P E-P o o ... o o [E]
> [s] o o o o o B-C I-P I-P E-P o o ... o o [E]
> ```
> 
> . We would have to watch lower-ranked paths. The second reason is that there are several cases where the same mention span is assigned multiple entity types ([link](https://github.com/yahshibu/nested-ner-tacl2020-transformers/issues/3)). In such cases, we cannot decide which tag sequence should be ranked higher. By using separate CRFs for different entity types, we don't need to consider which should be ranked higher. The third reason is that [Muis and Lu (2017)](https://www.aclweb.org/anthology/D17-1276/) demonstrated that this approach of multiple CRFs would perform better on nested NER datasets and even a flat NER dataset than the standard approach of a single CRF for all entity types.

for first reason,  in the example， 
 [s] o o o o o B-P I-P I-P E-P o o ... o o [E]
[s] o o o o o B-C I-P I-P E-P o o ... o o [E]
 B-C I-P should be considered as illegal transition by setting the transition matrix.    Why the 2nd best path in a single CRF could get result like this example ?

**Comment by yahshibu** [comment]
You correctly understand our algorithm, but I'm sorry that I'd mistyped in my previous comment and that I confused you.

I wanted to type as follows:
```
[s] o o o o o B-P I-P I-P E-P o o ... o o [E]
[s] o o o o o B-C I-C I-C E-C o o ... o o [E]
```
In the case where the same span is assigned two entity types P and C, we cannot decide which should be ranked higher, P or C.

Does it make sense?

**Comment by wangwang110** [comment]
> You correctly understand our algorithm, but I'm sorry that I'd mistyped in my previous comment and that I confused you.
> 
> I wanted to type as follows:
> 
> ```
> [s] o o o o o B-P I-P I-P E-P o o ... o o [E]
> [s] o o o o o B-C I-C I-C E-C o o ... o o [E]
> ```
> 
> In the case where the same span is assigned two entity types P and C, we cannot decide which should be ranked higher, P or C.
> 
> Does it make sense?

thank you for your patient， I got .  but as you explain， the first reason and the second reason are same

**Comment by yahshibu** [comment]
No problem. Questions are welcome.

Um, yes, the first and second reasons are the same. I split it into the two, but I've realized that they are the same. Thanks.
